### ISSUE TITLE
match media css screen

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/TorZillaPrint#45]
@kkapsner 

view-source:https://canvasblocker.kkapsner.de/test/screenSizeTest.js

Is there some other method of getting the screen (and/or inner window) here that I should add to the screen section?

Edit: test page link: https://canvasblocker.kkapsner.de/test/screenSizeTest.html

**Comment by KOLANICH** [comment]
BTW, why to use a recursion here instead of a loop?

**Comment by Thorin-Oakenpants** [comment]
So, I'm totally confused ... as per usual
- 2nd test: `addTest("media queries: window.matchMedia"`
- 3rd test: `addTest("media queries: css"`

---

So the 3rd test `media queries: css` is using JS to enumerate through the values and get a result: rather than use a css file? I'm struggling to follow your code @kkapsner . Anyway, I see one problem: it should be using `min-device-width` and `min-device-height` .. i.e with **min**

I like this, as an extra test to add (assuming it adds something new). I'll have to see how TF I can rework it so I understand it. And I want to time it, and make it more performant (I can see ways to do this, but it depends on whether or not I understood what it's actually doing). **If you feel like ELI5, that would be cool** (go easy on me)

----

The 2nd test `media queries: window.matchMedia` looks similar, and I have some other matchMedia tests. So this looks like something to add as well.

---

I'm just not sure if these are only duplicating the tests I already have: i.e the JS and matchMedia tests simply draws on the css @ media rules -> RFP spoof. It'll be interesting if blocking css blocks these. IDK, I mean I can block css in uMatrix, but that's just the css files, not inline via JS?.

When I work out WTF is going on and if I add them (I think there's something new here for interest's sake), I would also probably want them updated in real time as almost all the other screen and windows metrics are for desktop, which I can easily do.

**Comment by Thorin-Oakenpants** [comment]
![test](https://user-images.githubusercontent.com/16656956/68087727-3ca81d00-fe50-11e9-8afa-439d11072f5f.png)

**Comment by kkapsner** [comment]
@KOLANICH The recursion was a window.setTimeout recursion once (In developing the complete tab froze as I had the break condition wrong... so doing it asynchronously helped there). At the end I will streamline the test code as well and might go for a normal loop.

@Thorin-Oakenpants You could add to TPZ a check if all these methods return the same value. RFP seems to be fine but something like CB has a hard time to fake the 3rd (css) test. (It's like some other test pages likes https://canvasblocker.kkapsner.de/test/test.html where the result is basically always the same but the attack vector is different  - 

The code is just a binary search for the screen resolution (the 2nd and 3rd are basically the same but use different APIs to leverage the @media queries). I assume the screen dimension is between 0 and 2^32 (I think a quite fair assumption...) and the check the middle point between the min and max value if the actual device dimension is it or lower (so ...-min vs. ...-max does change the logic a little bit but is equivalent). Then adjust the range accordingly and repeat until you found the resolution.

**Comment by Thorin-Oakenpants** [comment]
Thanks. I think.... 
- the 3rd test is redundant, since we already have css files with those rules, and I want the results to show when JS is disabled - those two css files (screen and window = 471kb and are range limited to 500 - 1440 to save space) - but I might add something to prove a point
- the 2nd test is good to add, since it bypasses the css block

**Comment by Thorin-Oakenpants** [comment]
Test 2: media queries: window.matchMedia

Not sure how important `Math.pow(2, 32)` is. If I change to `Math.pow(2, 16)` it's approx three times faster, at 2ms :) But I'm probably missing **why** it needs to be `32`. The original test takes 6ms.

**Comment by Thorin-Oakenpants** [comment]
https://thorin-oakenpants.github.io/testing/screenmatchmedia.html

Not sure that I covered trapping all possible errors? @kkapsner . I won't add it to TZP until you finalize your test.

The resize event in TZP is limited to desktop (cuz reasons) but in my test above it's not. On my machine, seems the initial run is about 5ms, but a resize is `0` or `1` ms. Not sure if the perf timing is perfect: seems to be.

**Comment by Thorin-Oakenpants** [comment]
I should also be able to mod this with an extra parameter to get the inner window

**Comment by Thorin-Oakenpants** [comment]
With this matchMedia test, when the screen value is tied to the inner window (a la RFP), then it's variable - and goes whacky when you zoom: values start to drop off in ways they don't with the actual *min* only values in the TZP css files

**Comment by kkapsner** [comment]
I just went for 2^32 because it's the biggest integer JS can handle. But I think there is a better way to get these values. And the weird output with the zoom might be due to an off by one error in the binary search - I will refine my code.

**Comment by kkapsner** [comment]
Updated my code. 

PS: Funny enough I just found out that the screen properties also change with zoom if you have RFP not enabled.

**Comment by Thorin-Oakenpants** [comment]
I'll have a play (and update) later. I found that **my** test page didn't return anything on android, but yours did

> But I think there is a better way to get these values

So no more binary search?

**Comment by Thorin-Oakenpants** [comment]
I'm doing something wrong here

<details><summary>all-in-one html</summary><p>

```html
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<title>screen matchMedia</title>
<style>
body {background-color: white; color: black;}
</style>
</head>
<body>

<div><a href="https://canvasblocker.kkapsner.de/test/screenSizeTest.html" target="blank">kkapsner test</a></div>
<hr>
<div>[matchMedia] screen resolution: <span id="outputscreen"></span></div>
<div>[matchMedia] inner window: <span id="outputwindow"></span></div>
<div>perf: <span id="perf"></span></div>

<script>

function get_matchmedia() {
	// variables
	let w = "", h = "", minValue = 0, maxValue = 512;
	// run them
	function runTest(callback){
		let t0 = performance.now();

		//callback("width", "device-").then(function(value){
		//	w = value;
		//	console.debug("w", w);
		//});
		//callback("height", "device-").then(function(value){
		//	h = value;
		//	console.debug("h", h);
			//document.getElementById("outputscreen").innerHTML = w+" x "+h;
		//});

		//callback("width", "device-").then();
		//callback("height", "device-").then();

		//callback("width", "").then();
		callback("height", "").then();

		let t1 = performance.now();
		document.getElementById("perf").innerHTML = (t1-t0) + " ms";
	}

	function searchValue(tester){
		function stepUp(){
			return tester(maxValue).then(function(testResult){
				if (testResult === searchValue.isEqual){
					console.log("isEqual", maxValue); // this doesn't fire
					return maxValue;
				}
				else if (testResult === searchValue.isBigger){
					console.log("isBigger", maxValue);
					minValue = maxValue;
					maxValue *= 2;
					return stepUp();
				}
				else {
					console.log("isSmaller", maxValue);
					return false;
				}
			});
		}
		function binarySearch(){
			if (maxValue - minValue < 0.125){
				return tester(minValue).then(function(testResult){
					if (testResult.isEqual){
						return minValue;
					}
					else {
						return tester(maxValue).then(function(testResult){
							if (testResult.isEqual){
								return maxValue;
							}
							else {
								// we should round here and resolve
								return Promise.reject(
									"Between " + minValue + " and " + maxValue
								);
							}
						});
					}
				});
			}
			else {
				let pivot = (minValue + maxValue) / 2;
				return tester(pivot).then(function(testResult){
					if (testResult === searchValue.isEqual){
						console.debug(pivot); // found it
						return pivot;
					}
					else if (testResult === searchValue.isBigger){
						minValue = pivot;
						return binarySearch();
					}
					else {
						maxValue = pivot;
						return binarySearch();
					}
				});
			}
		}
		return stepUp().then(function(stepUpResult){
			if (stepUpResult){
				return stepUpResult;
			}
			else {
				return binarySearch();
			}
		});
	}
	searchValue.isSmaller = -1;
	searchValue.isEqual = 0;
	searchValue.isBigger = 1;

	runTest(function(type, metric){
		console.debug("running", type, metric);
		return searchValue(function(valueToTest){
			if (window.matchMedia("(" + metric + type + ": " + valueToTest + "px)").matches){
				return Promise.resolve(searchValue.isEqual);
			}
			else if (window.matchMedia("(max-" + metric + type + ": " + valueToTest + "px)").matches){
				return Promise.resolve(searchValue.isSmaller);
			}
			else {
				return Promise.resolve(searchValue.isBigger);
			}
		});
	});

};

get_matchmedia();
// window.addEventListener('resize', get_screen_matchmedia);

</script>
</body>
</html>
```

</p></details>

in the function `runTest(callback)` I can "call" any of the 4 results (width/height for screen/inner) I want and they work
e.g `callback("width", "device-").then();` the console outputs
```
running width device-
isBigger 512
isBigger 1024
isBigger 2048
isSmaller 4096
2560
```
but if I try to run more than one, I get an error: "uncaught exception: Between 2048 and 2048.0625" - I'm not sure what I'm missing here that your test works, but mine doesn't.

if I wait a bit (that is the first callback is completed) and reset the values, then it works
```js
		callback("width", "device-").then();
		setTimeout(function(){
			w = "", h = "", minValue = 0, maxValue = 512;
			callback("height", "device-").then();
		}, 100);
```

but, I don't want to wait: speed is of the essence. I can cascade it: but this is icky? And the perf gets up to 10ms which sucks donkey balls

```js
		callback("width", "device-").then(function(value){
			w = value;
			minValue = 0, maxValue = 512; // reset
			callback("height", "device-").then(function(value){
				h = value;
				document.getElementById("outputscreen").innerHTML = w+" x "+h;
			});
		});
```

I'm clearly missing something here

**Comment by kkapsner** [comment]
I still use a binary search, but the range to search in is got differently.

You are using the same variables (minValue and maxValue) for all tests and then they get mixed and confused. You have to put the variable definition inside searchValue.

The zoom confuses the media queries a little bit. max-device-... still work with integers but device-... itself requires you to end also enter the decimal part. I'm still working on my test page (uploaded the current version just now) and got rid of this issue by setting the search limit to 0.01 instead of 0.125.

**Comment by Thorin-Oakenpants** [comment]
I'm just going to finish a PoC for you, to show you that you need to reverse the mathematics and use min, not max. This solved the issues with no value being found (and you can ignore all the decimal shit). But this is only tested on Firefox, I don't care about other browsers

**Comment by kkapsner** [comment]
Why should it matter if I use min or max? In my test page I use both and they return the same value.

But I will have a look at your PoC.

**Comment by Thorin-Oakenpants** [comment]
OK, I got confused. I checked my followups issue which says

>  - css media window/screen: use `min-`width, `min-`height, `min-`device-width, `min-`device-height
>       *  without `min-` output is affected by zoom

Edit: I don't know why my [updated github file](https://github.com/Thorin-Oakenpants/testing/blob/master/screenmatchmedia.html) doesn't load [here](https://thorin-oakenpants.github.io/testing/screenmatchmedia.html) and keeps showing the old one (no it's not cache or anything: must be github dragging their ass somewhere) .. end edit

I think the "affected by zoom" refers to some values not showing. I guess it could have also used `max`
![max-min](https://user-images.githubusercontent.com/16656956/68216636-15308c00-ffd9-11e9-9490-1dd9a4331efe.png)

**Comment by Thorin-Oakenpants** [comment]
OK, that new one is smooth AF ... :beer:

**Comment by Thorin-Oakenpants** [comment]
OK, I don't know WTF is going on. Yours never misses any values. I'm doing the same thing, except concatenating the results for output. But I get missing values (not at 100% because resolutions are normally at integers at that res, I guess, IDK about DPI etc on mobile/retina): i.e, I get a blank sometimes depending on the zoom and original 100% zoom value

Plus, I'm also getting the inner window measurements, but that's just a change in the matchMedia argument. That too gets blanks.

I'll post the whole new test file and see if you can work it out (if you get time)

Also, see the bit (in my file) that says
```js
// we could round here and resolve
return Promise.reject(
	"between " + minValue + " and " + maxValue
);
```

I believe this should be trivial to round the number *down* to an integer and test it for a match, then add one and test again: you should get one of the two as a match: that pic two posts up always finds a match using whole numbers. I can look at this later, because I don't want to return decimals: as long as I'm consistent in using min or max, then everyone is treated the same and the entropy isn't affected

anyway .. next post on the way

**Comment by Thorin-Oakenpants** [comment]
this should be the same as yours?

<details><summary>all-in-one broken thorin matchMedia test</summary><p>

```html
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<title>matchMedia:Thorin:slightly broken</title>
<style>
body {background-color: white; color: black;}
</style>
</head>
<body>

<div><a href="https://canvasblocker.kkapsner.de/test/screenSizeTest.html" target="blank">kkapsner test</a></div>
<hr>
<div>[matchMedia] screen resolution: <span id="outputscreen"></span></div>
<div>[matchMedia] inner window: <span id="outputwindow"></span></div>
<div>perf: <span id="perf"></span></div>

<script>

function get_matchmedia() {
	'use strict';

	// run them
	function runTest(callback){
		let t0 = performance.now();
		// device
		let	wd = "", hd = ""
		callback("width", "device-").then(function(value){
			wd = value;
		});
		callback("height", "device-").then(function(value){
			hd = value;
			document.getElementById("outputscreen").innerHTML = wd+" x "+hd;
		});
		// inner window
		let	wi = "", hi = ""
		callback("width", "").then(function(value){
			wi = value;
		});
		callback("height", "").then(function(value){
			hi = value;
			document.getElementById("outputwindow").innerHTML = wi+" x "+hi;
			let t1 = performance.now();
			document.getElementById("perf").innerHTML = (t1-t0) + " ms";
		});
	}

	function searchValue(tester){
		var minValue = 0;
		var maxValue = 512;
		var ceiling = Math.pow(2, 32);
		
		function stepUp(){
			if (maxValue > ceiling){
				return Promise.reject("unable to find upper bound");
			}
			return tester(maxValue).then(function(testResult){
				if (testResult === searchValue.isEqual){
					return maxValue;
				}
				else if (testResult === searchValue.isBigger){
					//console.debug("isBigger", maxValue);
					minValue = maxValue;
					maxValue *= 2;
					return stepUp();
				}
				else {
					//console.debug("isSmaller", maxValue);
					return false;
				}
			});
		}
		function binarySearch(){
			if (maxValue - minValue < 0.01){
				return tester(minValue).then(function(testResult){
					if (testResult.isEqual){
						return minValue;
					}
					else {
						return tester(maxValue).then(function(testResult){
							if (testResult.isEqual){
								return maxValue;
							}
							else {
								// we could round here and resolve
								return Promise.reject(
									"between " + minValue + " and " + maxValue
								);
							}
						});
					}
				});
			}
			else {
				let pivot = (minValue + maxValue) / 2;
				return tester(pivot).then(function(testResult){
					if (testResult === searchValue.isEqual){
						//console.debug("found it", pivot); // found it
						return pivot;
					}
					else if (testResult === searchValue.isBigger){
						minValue = pivot;
						return binarySearch();
					}
					else {
						maxValue = pivot;
						return binarySearch();
					}
				});
			}
		}
		return stepUp().then(function(stepUpResult){
			if (stepUpResult){
				return stepUpResult;
			}
			else {
				return binarySearch();
			}
		});
	}
	searchValue.isSmaller = -1;
	searchValue.isEqual = 0;
	searchValue.isBigger = 1;

	runTest(function(type, metric){
		"use strict";
		//console.debug("running", type, metric);
		return searchValue(function(valueToTest){
			if (window.matchMedia("(" + metric + type + ": " + valueToTest + "px)").matches){
				return Promise.resolve(searchValue.isEqual);
			}
			else if (window.matchMedia("(max-" + metric + type + ": " + valueToTest + "px)").matches){
				return Promise.resolve(searchValue.isSmaller);
			}
			else {
				return Promise.resolve(searchValue.isBigger);
			}
		});
	});

};

get_matchmedia();
window.addEventListener('resize', get_matchmedia);

</script>
</body>
</html>
```

</p></details>

**Comment by kkapsner** [comment]
You have to treat the promises differently. It's not guaranteed that the promise for wd resolves before the wh. So you have to use Promise.all() to wait for both of them and then output both:```
	function runTest(callback){
		let t0 = performance.now();
		// device
		let devicePromise = Promise.all([
			callback("width", "device-"),
			callback("height", "device-")
		]);
		devicePromise.then(function(wd, hd){
			document.getElementById("outputscreen").innerHTML = wd+" x "+hd;
		});
		// inner window
		let innerPromise = Promise.all([
			callback("width", ""),
			callback("height", "")
		]);
		innerPromise.then(function(wi, hi){
			document.getElementById("outputwindow").innerHTML = wi+" x "+hi;
		});
		Promise.all([devicePromise, innerPromise]).then(function(){
			let t1 = performance.now();
			document.getElementById("perf").innerHTML = (t1-t0) + " ms";
		});
	}
``` - not tested.

But rounding to integer does not give you a match with device-... as you can see in my latest update of my test page (in the consistency checks).

**Comment by Thorin-Oakenpants** [comment]
Thanks... I'll give that a try. I need to read more about promises, I tired to a while ago to make one of the other tests return results as fast as possible but failed miserably after ~~half~~ a whole day of tinkering and reading

> rounding to integer does not give you a match with device-

My PoC (which is now gone but was just to test something, looping the whole thing sucks, but I have a local copy), **always works** (incl zooms) and **only uses integers**: try this (example min), using 7681 as the upper 8K screen limit

```js
	// this sucks donkey balls: ~100ms
	let sw="", sh="", iw="", ih="";
	for (let i = 7681; i > 1; i--) {
		if (sw == "") { if (matchMedia("(min-device-width:" + i + "px)").matches) {sw=i};}
		if (sh == "") { if (matchMedia("(min-device-height: " + i + "px)").matches) {sh=i}}
		if (iw == "") { if (matchMedia("(min-width:" + i + "px)").matches) {iw=i}}
		if (ih == "") {if (matchMedia("(min-height:" + i + "px)").matches) {ih=i}}
	}
	console.debug("screen", sw+" x "+sh);
	console.debug("screen", iw+" x "+ih);
```

**Comment by Thorin-Oakenpants** [comment]
> not tested

the promise returns a single result, so I had to `toString` and `.replace` it
```js
		devicePromise.then(function(device){
			device = device.toString().replace(",", " x ")
			document.getElementById("outputscreen").innerHTML = device;
		});
```

**Comment by Thorin-Oakenpants** [comment]
PS: i changed by screen and window css files [here](https://github.com/ghacksuserjs/TorZillaPrint/tree/master/css) to use `max-` and on android they return the max result in the file (2560, my range is 500 to 2560) .. I knew there was something about needing to use `min-` (at least in the css files, not necessarily your code).

**Comment by Thorin-Oakenpants** [comment]
Sweet .. all done, looking [awesome](https://ghacksuserjs.github.io/TorZillaPrint/TorZillaPrint.html#screen). I also rounded the pivot **down** when found, so it matches my css min rules ... and credited <sup>1</sup> you. Thanks sooooooo much 💋 + 🍻 

<sup>1</sup> I linked directly to the test page since your index doesn't link it yet : let me know when you add it

**question**: where I have credited you (domrect, etc), I have two links, one to your github account https://github.com/kkapsner, the other to githubs CB page https://github.com/kkapsner/CanvasBlocker. This seems a little redundant

Would you like me to change these to something else, e.g. the test index or AMO. And instead of kkapsner, should I use your real name (Korbinian Kapsner, right?) like all the others (except earthlng who wants to remain anonymous)

**Comment by Thorin-Oakenpants** [comment]
PPS: on android, your code re-calculates when scrolling (at least on my phone) - which is no big deal :) On TZP, I only use the on resize event for desktop (for a whole bunch of other reasons: e.g in order to grab keyboard height, toolbar height, etc)

**Comment by Thorin-Oakenpants** [comment]
> 
> 
> Updated my code.
> 
> PS: Funny enough I just found out that the screen properties also change with zoom if you have RFP not enabled.

They **all** do (inner, outer, available, screen, viewport ), open TZP in a normal FF, no RFP and zoom in and out

**Comment by kkapsner** [comment]
> the promise returns a single result, so I had to toString and .replace it
You're right - in that case it's an array. So you could simply to a ```device.joins("x")```.

I included the screen test into my overview page so you can reference to there.

Having both references seems also redundant to me. Just link to CB as the code is basically from this repo.

**Comment by Thorin-Oakenpants** [comment]
So ... after [this](https://old.reddit.com/r/TOR/comments/e57ow3/request_for_community_feedback_does_your_tor/) reddit post, I've made a couple of changes

First of all, I've added notation to the css field titles that they're using `min-`. Secondly, instead of rounding matchMedia screen/inner-window measures so they match the css, I now output the actual value, decimals and all - this should eliminate any confusion.

Entropy was already there, by design. i.e if DPI !=1 then on some measurements there can a pixel difference on height and/or width on the css+matchMedia vs the window.* and screen* tests. Now it's more precise. Since the matchMedia test is bisecting, maybe the final value will always end up the same - i.e between 1800 and 1801 it will always be 1800.296875? But I suspect it's more likely that different DPI values (different device specs) will result in different results - I'll add it to my list of things to investigate.

(yes: I got a new phone @KOLANICH )
![example](https://user-images.githubusercontent.com/16656956/70337641-6e434800-1843-11ea-9817-2ed0f73afef4.png)

**Comment by Thorin-Oakenpants** [comment]
@kkapsner ... quick question

when bisecting and a decimal is involved, does it stop at some point and just output a value, or does it go until it finds a match. I see code for the `promise.reject` - what triggers this exactly?

```
return Promise.reject(
	"between " + minValue + " and " + maxValue
);
```

Is the precison controlled by the `0.01` value, and if I ramped it up to `0.001` we could go nuts? I think I answered my own question - just not sure on the perf impact

Here's what I'm angling at: precision/increase of the decimal places. e.g my phone has a real DPI of `2.65` and I return `.296875` (at 1800/1801 pixels high, but I don't think the height matters) and I'm wondering if this can be used (not that I can see yet) to calculate the DPI - IDK. i.e we already have some entropy, but I wonder if a phone with e.g DPI `2.60` returns the same value etc

**Comment by kkapsner** [comment]
Yes, the precision is controlled by the 0.01. But as far as my tests go the media queries are handled by the same precision and increasing to 0.001 should not change anything. But you can try...

The DPI should already be exposed by `window.devicePixelRatio` so need to calculate it.

**Comment by Thorin-Oakenpants** [comment]
`window.devicePixelRatio` is spoofed by RFP as 1 (they're thinking about changing that to 2)

Yup, I changed https://thorin-oakenpants.github.io/testing/ to use `0.001` and no diff on my android (the only device I have with DPI != 1)

> the media queries are handled by the same precision

Do you mean matchMedia's precision can't be any more precise?

**Comment by kkapsner** [comment]
Ah... didn't know about RFP changing that. Kind of messes up the screen protection of CB...

> Do you mean matchMedia's precision can't be any more precise?

They were not in my tests.

**Comment by Thorin-Oakenpants** [comment]
:facepalm: :bashheadonwall: I keep saying `DPI` (e.g. when DPI !=1, which is stupid) .. and what I really meant was `devicePixelRatio` .. maybe I need more beer

> They were not in my tests

nvm, what I meant was that the implementation of the matchMedia API in firefox can't provide more precision: not your tests
